WorkloadManagementSystem/Utilities/QueueUtilities.py

```diff
@@ -0,0 +1,225 @@
+          valueToPrint = "%s..." % valueToPrint[:20]
+        noMatchReasons.append('Job %s %s requirement not satisfied' % (parameter, valueToPrint))
+        if not fullMatch:
+          return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # 4. Tags
+  tags = set(job.getListFromExpression('Tag'))
+  nProc = job.getAttributeInt('NumberOfProcessors')
+  if nProc and nProc > 1:
+    tags.add('MultiProcessor')
+  wholeNode = job.getAttributeString('WholeNode')
+  if wholeNode:
+    tags.add('WholeNode')
+  queueTags = set(queueDict.get('Tags', []))
+  if not tags.issubset(queueTags):
+    noMatchReasons.append('Job Tag %s not satisfied' % ','.join(tags))
+    if not fullMatch:
+      return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # 4. MultiProcessor requirements
+  if nProc and nProc > int(queueDict.get('NumberOfProcessors', 1)):
+    noMatchReasons.append('Job NumberOfProcessors %d requirement not satisfied' % nProc)
+    if not fullMatch:
+      return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # 5. RAM
+  ram = job.getAttributeInt('RAM')
+  if ram and ram > int(queueDict['MaxRAM']):
+    noMatchReasons.append('Job RAM %d requirement not satisfied' % ram)
+    if not fullMatch:
+      return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # Check resource requirements to job
+  # 1. OwnerGroup - rare case but still
+  if "OwnerGroup" in queueDict:
+    result = getProxyInfo(disableVOMS=True)
+    if not result['OK']:
+      return S_ERROR('No valid proxy available')
+    ownerGroup = result['Value']['group']
+    if ownerGroup != queueDict['OwnerGroup']:
+      noMatchReasons.append('Resource OwnerGroup %s requirement not satisfied' % queueDict['OwnerGroup'])
+      if not fullMatch:
+        return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # 2. Required tags
+  requiredTags = set(queueDict.get('RequiredTags', []))
+  if not requiredTags.issubset(tags):
+    noMatchReasons.append('Resource RequiredTags %s not satisfied' % ','.join(requiredTags))
+    if not fullMatch:
+      return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  # 3. RunningLimit
+  site = queueDict['Site']
+  opsHelper = Operations()
+  result = opsHelper.getSections('JobScheduling/RunningLimit')
+  if result['OK'] and site in result['Value']:
+    result = opsHelper.getSections('JobScheduling/RunningLimit/%s' % site)
+    if result['OK']:
+      for parameter in result['Value']:
+        value = job.getAttributeString(parameter)
+        if value and opsHelper.getValue('JobScheduling/RunningLimit/%s/%s/%s' % (site, parameter, value), 1) == 0:
+          noMatchReasons.append('Resource operational %s requirement not satisfied' % parameter)
+          if not fullMatch:
+            return S_OK({'Match': False, 'Reason': noMatchReasons[0]})
+
+  if noMatchReasons:
+    return S_OK({'Match': False, 'Reason': noMatchReasons})
+  return S_OK({'Match': True, 'Reason': noMatchReasons})
```

Comment:
I think this is better rather than having two branches to test:
```suggestion
  return S_OK({'Match': not bool(noMatchReasons), 'Reason': noMatchReasons})
```

Reply:
Lines like that are certainly smarter. But when reading the code you will certainly spend more time to read and understand this one line than reading the three lines above. Squeezing several lines of code in just one is not always better (although smarter :) )

Reply:
Would you find something like this better?
```suggestion
  hasMatch = len(noMatchReasons) == 0
  return S_OK({'Match': hasMatch, 'Reason': noMatchReasons})
```
The reason I suggest avoiding the `if` is to avoid accidentally making mistakes like below. In this case I'm almost it's correct but reducing the number of "paths" through the code makes it much easier to test and maintain.
```python
  if noMatchReasons:
    return S_OK({'Match': False, 'Reeson': noMatchReasons})
  return S_OK({'Match': True, 'Reason': noMatchReasons})
```